### PR TITLE
Integration testing app.js

### PR DESCRIPTION
**Pull request by stephenisau**
This pull request includes a

- [ ] Bug fix
- [X] New Feature
- [X] Testing
- [ ] Translation

**The following changes were made:**

- Added supertest package to package.json
- Modified package.json scripts
- Added __mocks__ folder to mock methods built-in to the fs module (specifically fs.readdirSync, fs.readFileSync to return custom data)
- Added route and integration tests to `__tests__/integration/app.test.js`
- Added strings around mode value on L11 in `routes/api.js` to fix legacy octal error in strict mode.

**If this is related to an existing ticket, include a link to it as well.**  
Issue [#5](https://github.com/rithmschool/numbers_api/issues/5)

### CODE REVIEW

**Comment by mmmaaatttttt** [comment]
@stephenisau Now that #33 is merged, can you please update this PR with the latest changes from master? should help clean up the diff. looks like there will be some merge conflicts, not sure why.

**Comment by mmmaaatttttt** [comment]
Hey @stephenisau -- i'm a little confused because i thought the diff would be smaller, but it doesn't really seem to be. For example, in `master` there is now a `__mocks__/fs.js` file, but this PR shows that file as being added. Any idea what's going on here? if not, maybe it's worth closing the PR and reopening a fresh one off master? if that doesn't fix the large diff, maybe we can hop on a call to try to figure out what's going on.

**Comment by stephenisau** [comment]
@mmmaaatttttt that's kinda weird. I originally had to move `__mocks__/fs.js` into this branch as a commit because I had messed up earlier with branching. I can close this PR out and re-open a new one off master^

**Comment by stephenisau** [comment]
@mmmaaatttttt Actually, would you mind checking if the diff has been fixed? I was able to pull and merge the changes from `rithm:master` into this branch where the differences should be fixed now. If it's still an issue, I can close and re-open a new PR!

**Comment by mmmaaatttttt** [comment]
@stephenisau - diff looks great now! i'll review.

File: __tests__/integration/app.test.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+const api = require("../../routes/api");
+
+describe("Integration/unit testing app.js", function () {
+  describe("GET /", function () {
+    test("GET request responds with a 200 status code and returns HTML", async function (done) {
+      const response = await request(app).get("/");
+      const { statusCode, text } = response;
+      expect(statusCode).toBe(200);
+      expect(text).toContain("<!DOCTYPE html>");
+      done();
```
Comment (mmmaaatttttt) [inline __tests__/integration/app.test.js:13]: hmm, do you need this `done` pattern? i'd try removing it and seeing if the tests still work.

File: __tests__/integration/app.test.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+const fact = require("../../models/fact");
+const api = require("../../routes/api");
+
+describe("Integration/unit testing app.js", function () {
+  describe("GET /", function () {
+    test("GET request responds with a 200 status code and returns HTML", async function (done) {
+      const response = await request(app).get("/");
+      const { statusCode, text } = response;
+      expect(statusCode).toBe(200);
+      expect(text).toContain("<!DOCTYPE html>");
```
Comment (mmmaaatttttt) [inline __tests__/integration/app.test.js:12]: can you get a bit more specific about what's returned? like maybe find something in the html that's specific to the documentation?

File: __tests__/integration/app.test.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      expect(getFact).toHaveBeenCalled();
+      done();
+    });
+  });
+
+  describe("POST => /submit", function () {
+    test("POST request should respond with a 200 status code", async function (done) {
+      const response = await request(app).post("/submit");
+      const { statusCode } = response;
+      expect(statusCode).toBe(200);
```
Comment (mmmaaatttttt) [inline __tests__/integration/app.test.js:28]: can you also make an assertion on the body? is there anything else in the response here to help clarify what it looks like?

File: __tests__/integration/app.test.js (hunk 1 above)
Comment (mmmaaatttttt) [reply __tests__/integration/app.test.js:13]: (this comment applies to all tests in this suite)

File: package-lock.json [hunk 4]
```diff
@@ -6234,6 +6238,42 @@
...
+          "version": "3.2.6",
+          "resolved": "https://registry.npmjs.org/debug/-/debug-3.2.6.tgz",
+          "integrity": "sha512-mel+jf7nrtEl5Pn1Qx46zARXKDpBbvzezse7p7LqINmdoIk8PYP5SySaxEmYv6TZ0JyEKA1hsCId6DIhgITtWQ==",
+          "requires": {
+            "ms": "^2.1.1"
+          }
+        }
+      }
+    },
+    "supertest": {
```
Comment (mmmaaatttttt) [inline package-lock.json:6268]: hmm, i see supertest in here, but why isn't it also a `devDependency` in the `package.json`?

File: package.json [hunk 5]
```diff
@@ -2,8 +2,8 @@
   "name": "numbers_api",
   "version": "1.0.0",
   "scripts": {
-    "test": "jest",
-    "start": "NODE_ENV=development nodemon app.js",
+    "test": "jest --forceExit",
+    "start": "NODE_ENV=development node app.js",
```
Comment (mmmaaatttttt) [inline package.json:6]: do we need these changes?

File: routes/api.js [hunk 6]
```diff
@@ -8,7 +8,7 @@ exports.appendToFile = function (filePath, dataStr) {
   var stream = fs.createWriteStream(filePath, {
     flags: "a",
     encoding: "utf8",
-    mode: 0666,
+    mode: "0666",
```
Comment (mmmaaatttttt) [inline routes/api.js:11]: what's the reason for this change?

**Review by mmmaaatttttt — CHANGES_REQUESTED** [review]
@stephenisau - reviewed! lmk if you have any questions

File: routes/api.js (hunk 6 above)
Comment (stephenisau) [reply routes/api.js:11]: Weird... The reason for the change originally was because there was a warning about legacy octals not being supported, but I'm not seeing it when I hit the route.

I referenced this stackoverflow post as a fix - https://stackoverflow.com/questions/23609042/how-to-avoid-octal-literals-are-not-allowed-in-strict-mode-with-createwritestr

Comment (stephenisau) [reply routes/api.js:11]: So running `jest __tests__/integration/app.test.js` reproduces the legacy octal error I mentioned up here and the fix is to wrap the value in a string.

File: __tests__/integration/app.test.js (hunk 1 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:13]: removing the done() callbacks gives this error message for each of the tests - 
`    : Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error:
`

File: package-lock.json (hunk 4 above)
Comment (stephenisau) [reply package-lock.json:6268]: added this as a dev dependency!

File: routes/api.js (hunk 6 above)
Comment (stephenisau) [reply routes/api.js:11]: nevermind, I just fixed this by implementing the changes from [#35](https://github.com/rithmschool/numbers_api/pull/35) to this pull request. I'll go ahead and close #35 out @mmmaaatttttt!

File: __tests__/integration/app.test.js (hunk 3 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:28]: This one's a bit tricky. The post request for this route takes a request body and is supposed to return the same request body back to the user, but it's not enabled for the application yet. I just wrote this test to expect a 200 response code but I can make the changes if needed!

File: __tests__/integration/app.test.js (hunk 2 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js]: I changed this to look for something more specific

File: __tests__/integration/app.test.js (hunk 3 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:29]: I've enabled json for the express app, and changed the test!

File: package.json (hunk 5 above)
Comment (stephenisau) [reply package.json:6]: Removed the `--forceExit` flag. I changed the `start` script to use `node` instead of `forever`.

Comment (stephenisau) [reply package.json]: Closing ticket #35 and implementing the changes in this ticket. That fixes the `--forceExit` issue.

File: __tests__/integration/app.test.js (hunk 1 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:14]: @mmmaaatttttt removing the `done()` pattern causes the tests to fail. I'm not sure there's another way to fix it without manually configuring jest.

Comment (mmmaaatttttt) [reply __tests__/integration/app.test.js:14]: Hey @stephenisau - I can take out these `done` references and everything works fine. What version of node are you using? I'm on 12.2.0.

Comment (mmmaaatttttt) [reply __tests__/integration/app.test.js:14]: Somewhat relatedly, I'm not sure if this is coming from the `app` tests or the `data` tests, but when I run the tests, my console gets hammered by warnings and errors. can you please make sure that both the tests pass, and the output doesn't contain warnings / errors? you should be able to mock console.warn and error and replace them with implementations that don't actually log anything. see here for some ideas: https://medium.com/swlh/6-ways-to-run-jest-test-cases-silently-67d2fead8c11

File: __tests__/integration/app.test.js (hunk 3 above)
Comment (mmmaaatttttt) [reply __tests__/integration/app.test.js:29]: maybe just get rid of this test then, i think the test below is sufficient.

File: __tests__/integration/app.test.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      done();
+    });
+
+    test("POST /submit request should not return data to user if no headers are set", async function (done) {
```
Comment (mmmaaatttttt) [inline __tests__/integration/app.test.js:33]: is this actually what you're testing? it seems like it returns data to the user even without setting any headers.

File: package.json [hunk 8]
```diff
@@ -3,7 +3,7 @@
   "version": "1.0.0",
   "scripts": {
     "test": "jest",
-    "start": "NODE_ENV=development nodemon app.js",
+    "start": "NODE_ENV=development node app.js",
```
Comment (mmmaaatttttt) [inline package.json:6]: any reason why you changed this from `nodemon` to `node`?

**Review by mmmaaatttttt — CHANGES_REQUESTED** [review]
@stephenisau just a few more questions on the tests, but this is close!

File: package.json (hunk 8 above)
Comment (stephenisau) [reply package.json:6]: No reason, I thought `node` was being used in the `package.json` on github. I should've checked before pushing this one 👎

File: __tests__/integration/app.test.js (hunk 3 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:29]: okay, I just got rid of this test. Should I disable json for the application as well? @mmmaaatttttt

File: __tests__/integration/app.test.js (hunk 1 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:14]: yeppers, i've added some jest configurations and a `setup.js` file that runs before each test suite. Referenced the article you've provided and left it as a comment in the `setup.js` file!

File: __tests__/integration/app.test.js (hunk 3 above)
Comment (mmmaaatttttt) [reply __tests__/integration/app.test.js:29]: @stephenisau  - i guess if you don't need it anymore, might as well remove.

File: __tests__/integration/app.test.js (hunk 7 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js:33]: ah, so if I send this in as part of the request body without headers, the route returns an empty object. But if I _do_ send in the request with header `Content-Type: Application/json`, along with json enabled in the app, then this _does_ work. 

I'm not really sure what the best option would be :|

Comment (mmmaaatttttt) [reply __tests__/integration/app.test.js:33]: hmm, as long as using the json middleware doesn't impact other routes, maybe the easiest thing is to just include it then?

File: __tests__/integration/app.test.js (hunk 3 above)
Comment (stephenisau) [reply __tests__/integration/app.test.js]: got it!

File: __tests__/integration/app.test.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    test("GET request should call fact.getFact()", async () => {
+      const getFact = jest.spyOn(fact, "getFact");
+      const response = await request(app).get("/");
+      expect(getFact).toHaveBeenCalled();
+    });
+  });
+
+  describe("POST => /submit", function () {
+    test("POST /submit request should not return data to user if no headers are set", async () => {
```
Comment (mmmaaatttttt) [inline __tests__/integration/app.test.js:24]:
```suggestion
    test("POST /submit request should return data to user", async () => {
```

File: __tests__/setup.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+global.console = {
+  ...console,
+  log: jest.fn(),
+  debug: jest.fn(),
+};
+
+jest.spyOn(global.console, "log").mockImplementation(jest.fn());
+jest.spyOn(global.console, "warn").mockImplementation(jest.fn());
+jest.spyOn(global.console, "error").mockImplementation(jest.fn());
```
Comment (mmmaaatttttt) [inline __tests__/setup.js:12]: let's keep `console.log` as-is, since it's often helpful to log things in the tests. i also don't think you need to use both of these patterns. does this still work if the contents are just:
```suggestion
// Configuring console messages to run silently while testing with jest
// https://medium.com/swlh/6-ways-to-run-jest-test-cases-silently-67d2fead8c11

global.console = {
  ...console,
  warn: jest.fn(),
  error: jest.fn(),
};
```

**Review by mmmaaatttttt — COMMENTED** [review]
just one comment on the console mocks! (you may need to pull down from GH since i pushed up a commit)

**Review by mmmaaatttttt — APPROVED** [review]
